juniper_subscriptions/src/lib.rs

```diff
@@ -135,6 +135,121 @@ where
+        #[serde(rename = "connection_terminate")]
+        GqlConnectionTerminate,
+        /// Client -> Server
+        /// Client sends this message to execute GraphQL operation
+        #[serde(rename = "start")]
+        GqlStart,
+        /// Server -> Client
+        /// The server sends this message to transfer the GraphQL execution result from the
+        /// server to the client, this message is a response for GQL_START message.
+        #[serde(rename = "data")]
+        GqlData,
+        /// Server -> Client
+        /// Server sends this message upon a failing operation, before the GraphQL execution,
+        /// usually due to GraphQL validation errors (resolver errors are part of GQL_DATA message, and will be added as errors array)
+        #[serde(rename = "error")]
+        GqlError,
+        /// Server -> Client
+        /// Server sends this message to indicate that a GraphQL operation is done,
+        /// and no more data will arrive for the specific operation.
+        #[serde(rename = "complete")]
+        GqlComplete,
+        /// Client -> Server
+        /// Client sends this message in order to stop a running GraphQL operation execution
+        /// (for example: unsubscribe)
+        #[serde(rename = "stop")]
+        GqlStop,
+    }
+
+    /// Subscription Protocol Message Type Values over WS
+    #[allow(missing_docs)]
+    pub mod message_types {
+        pub const GQL_CONNECTION_INIT: &str = "connection_init";
+        pub const GQL_CONNECTION_ACK: &str = "connection_ack";
+        pub const GQL_CONNECTION_ERROR: &str = "connection_error";
+        pub const GQL_CONNECTION_KEEP_ALIVE: &str = "ka";
+        pub const GQL_CONNECTION_TERMINATE: &str = "connection_terminate";
+        pub const GQL_START: &str = "start";
+        pub const GQL_DATA: &str = "data";
+        pub const GQL_ERROR: &str = "error";
+        pub const GQL_COMPLETE: &str = "complete";
+        pub const GQL_STOP: &str = "stop";
+    }
+
+    /// Empty SubscriptionLifeCycleHandler over WS
+    pub enum SubscriptionState<'a, Context>
+    where
+        Context: Clone + Send + Sync,
+    {
+        /// The Subscription is at the init of the connection with the client after the
+        /// server receives the GQL_CONNECTION_INIT message.
+        OnConnection(Option<String>, &'a mut Context),
+        /// The Subscription is at the start of a operation after the GQL_START message is
+        /// is received.
+        OnOperation(&'a mut Context),
+        /// The subscription is on the end of a operation before sending the GQL_COMPLETE
+        /// message to the client.
+        OnOperationComplete(&'a Context),
+        /// The Subscription is terminating the connection with the client.
+        OnDisconnect(&'a Context),
+    }
+
+    /// Trait based on the SubscriptionServer
+    /// https://www.apollographql.com/docs/graphql-subscriptions/lifecycle-events/
+    pub trait SubscriptionStateHandler<Context, E>
```

Comment:
So i was considering this for authorization or log purposes into the subscription, since the websocket protocol doesnt handle authorization the same way as the http does. I saw that in the apollo client it seems that the authorization token or user information is sent as a payload in the ConnectionInit message, so i was trying to think in something that could enable the insertion of user information into the GraphQL Context or something that could enable easier ways to do it.